### DISCUSSION TITLE
PinOK returns to Item 1 instead of Selecting an item, when using a port expander

### DISCUSSION THREAD

**Discussion by alextrical** [davetcc/tcMenu#365] [Q&A] [answered]
I'm currently trying to use a Rotary encoder attached to a 8574 I2C port expander, however it now seems to have an odd issue whereby the OK button no-longer selects a menu entry or enters a sub menu. 
Instead it returns to menu item "ID 1" from wherever the cursor was previously on the screen.

I'm not really sure how to resolve this issue, please can you advise where I should look to start resolving this?

[PolyPen.zip](https://github.com/davetcc/tcMenu/files/11666908/PolyPen.zip)

**Comment by davetcc** [comment, marked as answer]
I think I would start with what versions of everything are you running. What changed between it working and not. If it has never worked, how does it differ from a working example. Does the basic example work on that hardware setup?

Enable serial logging within IoAbstraction (see documentation) and look at the logging output.

**Reply by alextrical** [reply]
Good point, sorry it slipped my mind to include the basic information
The setup above was built using the Menu 4.0.0 designer from a blank start.
Compiled using VSCode Platformio using the following library versions
	davetcc/tcMenu@4.0.0
	davetcc/IoAbstraction@4.0.0
	davetcc/tcMenu@4.0.0
	davetcc/tcUnicodeHelper@1.0.0
	olikraus/U8g2@2.34.18
	teemuatlut/TMCStepper@0.7.3
	SoftwareSerial@1.0.0
	
I haven't seen the encoder on an Port expander working when built through TCMenu designer (but it happily works when connected to the MCU's IO's)

The working example I have for this setup, Is based around the example provided with IoAbstraction (excluding all additional functionality)
```
#include <Wire.h>
#define SDAPin          PB4
#define SCLPin          PB3

#include<IoAbstraction.h>
#include<IoAbstractionWire.h>
#include <TaskManagerIO.h>

PCF8574IoAbstraction io8574(0x38, /*3*/25);

// The pin onto which we connected the rotary encoders switch
const int encoderSWPin = 5;

// The two pins where we connected the A and B pins of the encoder. I recommend you dont change these
// as the pin must support interrupts.
const int encoderAPin = 6;
const int encoderBPin = 7;

// the maximum value (starting from 0) that we want the encoder to represent. This range is inclusive.
const int maximumEncoderValue = 255;

//
// When the SW is clicked and released, the following two functions will be run
//
void onSWClicked(uint8_t /*pin*/, bool heldDown) {
    Serial.print("Encoder button pressed ");
    Serial.println(heldDown ? "Held" : "Pressed");
}

void onSWButtonReleased(uint8_t /*pin*/, bool heldDown) {
    Serial.print("Encoder released - previously ");
    Serial.println(heldDown ? "Held" : "Pressed");
}

class MyEncoderChangeListener : public EncoderListener {
public:
    void encoderHasChanged(int newValue) override {
        Serial.println(newValue);
    }
} encoderChangeListener;

void setup() {
    // Before doing anything else, we must initialise the wire and serial libraries, as we are using both.
    Serial.begin(9600);
    while(!Serial);
    Wire.setSDA(SDAPin);
    Wire.setSCL(SCLPin);
    Wire.begin();

    Serial.println("Starting interrupt switch PCF8574 example now");

    // First we set up the switches library, giving it the task manager and tell it where the pins are located
    // We could also of chosen IO through an i2c device that supports interrupts.
    // the second parameter is a flag to use pull up switching, (true is pull up).
    switches.initialiseInterrupt(asIoRef(io8574), true);

    // now we add the switches, we dont want the spin wheel button to repeat, so leave off the last parameter
    // which is the repeat interval (millis / 20 basically) Repeat button does repeat as we can see.
    switches.addSwitch(encoderSWPin, onSWClicked);
    switches.onRelease(encoderSWPin, onSWButtonReleased);

    // now we set up the rotary encoder, first we give the A pin and the B pin.
    // we give the encoder a max value of 255, always minimum of 0.
    setupRotaryEncoderWithInterrupt(encoderAPin, encoderBPin, &encoderChangeListener);
    switches.changeEncoderPrecision(0, maximumEncoderValue, 128, false, 1);
}

void loop() {
    taskManager.runLoop();
}
```

Ah, I hadn't considered that was an option, but its a great place for me to start tracking down the issue. Thank you

**Reply by davetcc** [reply]
OK so the above sketch works when you run it standalone, but the tcMenu generated example doesn't work?

**Reply by alextrical** [reply]
Yes, the sketch above works (with the possible exception that it takes 2 rotary clicks before the state change is shown)
The switch handling logic is working cleanly, showing a single click or held state

**Reply by alextrical** [reply]
Ive got one of the TCMenu examples up and working today, the SW is working as expected. So good news, it looks to be isolated to the config I was using? I believe this is a non issue for the library/menu system

If i can isolate the cause of the issue, I will let you know. But for now I believe this can be considered closed (as a result of user error)
(Now I need to figure out what I've broken ;) )

**Reply by alextrical** [reply]
So the mistake I made was entirely user error. I had defined the Rotary encoder twice while doing a test, once in the main INO and a second in the Menu designer (as I show left it enabled) no idea why that resulted in return to Item1 of the menu, but all good now 😅

Thank you for your patience and support. Could have sworn I tested a few different things before asking for help...


The intent being to get the software to dynamically change based upon the hardware connected. If no port expander present on the I2C bus then use the the MCU pins, but if it is present use the expander instead. (Bit of a crazy setup, but the intent is to keep support for legacy users who already have a working setup, but also support new HW setups)
